src/Dictionaries/TrieDictionary.cpp

```diff
@@ -31,20 +24,175 @@ namespace ErrorCodes
+            : addr(addr_)
+            , prefix(prefix_)
+            , row(row_)
+            , isv6(addr.family() == Poco::Net::IPAddress::IPv6)
+        {
+        }
+
+        const uint8_t * asIPv6Binary(uint8_t * buf) const
+        {
+            if (isv6)
+                return reinterpret_cast<const uint8_t *>(addr.addr());
+            memset(buf, 0, 10);
+            buf[10] = '\xFF';
+            buf[11] = '\xFF';
+            memcpy(&buf[12], addr.addr(), 4);
+
+            return buf;
+        }
+
+        inline UInt8 prefixIPv6() const
+        {
+            return isv6 ? prefix : prefix + 96;
+        }
+    };
+
+    struct IPv4Subnet
+    {
+        UInt32 addr;
+        UInt8 prefix;
+    };
+
+    struct IPv6Subnet
+    {
+        const uint8_t * addr;
+        UInt8 prefix;
+    };
 }
 
 static void validateKeyTypes(const DataTypes & key_types)
 {
-    if (key_types.size() != 1)
-        throw Exception{"Expected a single IP address", ErrorCodes::TYPE_MISMATCH};
+    if (key_types.empty() || key_types.size() > 2)
+        throw Exception{"Expected a single IP address or IP with mask", ErrorCodes::TYPE_MISMATCH};
 
     const auto & actual_type = key_types[0]->getName();
-
     if (actual_type != "UInt32" && actual_type != "FixedString(16)")
         throw Exception{"Key does not match, expected either UInt32 or FixedString(16)", ErrorCodes::TYPE_MISMATCH};
+
+    if (key_types.size() > 1)
+    {
+        const auto * mask_col_type = typeid_cast<const DataTypeUInt8 *>(key_types[1].get());
+        if (mask_col_type == nullptr)
+            throw Exception{"Mask do not match, expected UInt8", ErrorCodes::TYPE_MISMATCH};
+    }
+}
+
+template <typename T, typename Comp>
+size_t sort_and_unique(std::vector<T> & vec, Comp comp)
+{
+    std::sort(vec.begin(), vec.end(),
+              [&](const auto & a, const auto & b) { return comp(a, b) < 0; });
+
+    auto new_end = std::unique(vec.begin(), vec.end(),
+                               [&](const auto & a, const auto & b) { return comp(a, b) == 0; });
+    if (new_end != vec.end())
+    {
+        vec.erase(new_end, vec.end());
+        return std::distance(new_end, vec.end());
+    }
+    return 0;
+}
+
+template <typename T>
+static inline int compareTo(T a, T b)
+{
+    return a > b ? 1 : (a < b ? -1 : 0);
+}
+
+inline static UInt32 IPv4AsUInt32(const void * addr)
+{
+    return Poco::ByteOrder::fromNetwork(*reinterpret_cast<const UInt32 *>(addr));
+}
+
+/// Convert mapped IPv6 to IPv4 if possible
+inline static UInt32 mappedIPv4ToBinary(const uint8_t * addr, bool & success)
+{
+    success = addr[0] == 0x0 && addr[1] == 0x0 &&
+              addr[2] == 0x0 && addr[3] == 0x0 &&
+              addr[4] == 0x0 && addr[5] == 0x0 &&
+              addr[6] == 0x0 && addr[7] == 0x0 &&
+              addr[8] == 0x0 && addr[9] == 0x0 &&
+              addr[10] == 0xff && addr[11] == 0xff;
+    if (!success)
+        return 0;
+    return IPv4AsUInt32(&addr[12]);
```

Comment:
I think strict aliasing rule is violated here, cause we have ptr which points to `uint8_t` and `UInt32` (inside `IPv4AsUInt32`) at the same time.
As I remember, it is ok to use `char` instead of `uint8_t` here.

Reply:
Also, that's how IPv4ToIPv6 is implemented in another place:
https://github.com/ClickHouse/ClickHouse/blob/a0da588f0608a91240bf1136d955b81d7614fcde/src/Functions/FunctionsCoding.h#L489-L493

I think we just need to use unalignedStore/unalignedLoad functions

Reply:
Hm, I was wrong about strict aliasing. uint8_t is a typedef for unsigned char, so it is ok. (have mixed it with char8_t)

Reply:
So, can we leave current implementation? This function different from `mapIPv4ToIPv6` (I have `mapIPv4ToIPv6` in this file too, but in other place).

Reply:
yes